### PR TITLE
Issue #35 / flipNormal()

### LINKED ISSUE #35: flipNormal()
By default when creating a geometric object, its normals point outward of the object. For some cases it is useful to force the normals to point inside the object. Thus there is a need for a `void flipNormals(void)` method.

### PR DESCRIPTION
**Pull request by otmanesabir**
As mentioned in the previous PR, we'll be splitting this task into two. This is the first part.

Partially Resolves #35

### CODE REVIEW

File: modules/core/IPrim.h [hunk 1]
```diff
@@ -99,10 +81,41 @@ namespace rt {
...
+        * @param ray Ray intersecting the primitive
+        * @retunrn The normalized normal of the primitive at the ray - primitive intersection point
+        */
+        DllExport virtual Vec3f				doGetShadingNormal(const Ray& ray) const { return doGetNormal(ray); }
 	private:
-		const ptr_shader_t	m_pShader;			///< Pointer to the sahder, see @ref  IShader
-		std::string			m_name;				///< Optional name of the primitive
+		const ptr_shader_t	m_pShader;			///< Pointer to the shader, see @ref  IShader.
+		std::string			m_name;				///< Optional name of the primitive.
+		bool			    m_flipped;			///< Flag which helps decide whether to flip the normal or not.
```
Comment (ereator) [inline modules/core/IPrim.h]: We should not forget to initialize it to _false_ by defaut
`bool m_flipped = false;`

**Review by ereator — CHANGES_REQUESTED** [review]
Thank you so much for nice and clear implementation.
I have started checking yesterday the new functionality in this PR and found two problems, that we did not foreseen:
1. The inverting normals of a Solid does not make any effect. This is possible to fix in a very straightforward way, so I would not say it is a _big_ problem
2. The _larger_ problem is that inversion of the normals of a composite does not work as well and I do not know at the moment how to fix it. As the method `CCompositeGeometry::doGetNormal()` should not be ever called, flipping normals of a composite will never take effect. The brute-force idea to fix that is to make method `virtual void IPrim::flipNormals() = 0;` to be pure virtual and force all derived classes to implement it. Any better idea is highly appreciated...

**Comment by ereator** [comment]
I have fixed the first issue, I mentioned above and made the flipped flag to be false on initialisation. I have also added the implementation of the `flipNormal()` method to the CPrimComposite. The preliminary implementation is just the same as for the solids - flipping the normals of all the primitives in `m_vPrims1` and `m_vPrims2`. However, the result of _e.g._ substraction becomes incorrect. The logic of entering and exiting the primitives is affected, thus flipping normals on composite does not only flips the normals, but also changes the entire result. We need to think how to fix that.

**Comment by otmanesabir** [comment]
thank you for fixing the first issue!
However, I still believe flipping the composite probably won't fix this issue.
I have an idea on how to potentially fix this issue but first I can try to explain what the problem is.

When we flip a primitive in the CSG tree, we do it 'globally'. Therefore, on the next ray intersection, that composite is still flipped. This is not too bad; however, it causes a problem on the next ray evaluation. Basically, when we want to evaluate if a ray is "entering" or "exiting" the primitive, we rely on the normal to figure that out. However, since we flip the normal on the previous ray, then this evaluation is done wrongly because now the `getNormal(..)` method returns the inverted normal. This is not good. Because all subsequent evaluations are also messed up. I think we can have two solutions:

1. Call `doGetNormal()` in `CCompositeGeometry` ray evaluation to get the **original** normal instead this flipped one 
2. Create a getter `isFlipped()` and factor the result of this boolean when we're doing the ray evaluation. 

I believe this should fix this issue but could still cause artifacts when dealing with larger primitives:

Let's imagine a case where a prim sphere is doing union, intersection, and difference. The difference operation is first and flips the normal of this entire sphere and now the union and intersection operations will run into consideration that this entire solid is pointing differently to where it's supposed to be. Therefore, generally speaking, I'm not sure if adding this on a primitive level is the solution for removing the Dummy primitives.

**Comment by ereator** [comment]
I have tried out the idea number 2 from your comment and it gave a correct results on the earth scene! 
I have implemented it in a fast manner, so without getter `isFlipped()` but by introducing another `bool m_isFlipped = false` in the composite class. 
I will try to perform additional tests and if everything now running correctly, I believe we can clean-up the code and merge the PR :-)

**Comment by otmanesabir** [comment]
very nice! I'm very glad it worked! 
I would be happy to also further test this over the coming weekend and if everything works well, can also clean up and attach a full description with my test scenes :D

**Comment by otmanesabir** [comment]
Implemented this on the primitive level and cleaned up the implementation a bit. I also got rid of this unused PrimDummy now :)


Attached is a scene I tested with also multiple operations. (Primitives and Solids). I think all looks good now but I could've missed something so would also wait for your confirmation. We can also think of adding unit tests for this composite logic to make sure that no changes break it. :) 


Screenshot 2021-12-04 at 12 07 48

File: modules/core/Ray.h [hunk 2]
```diff
@@ -18,7 +18,7 @@ namespace rt {
 		size_t 							counter;											///< Number of re-traces
 		
 		double							t		= std::numeric_limits<double>::infinity();	///< Current/maximum hit distance
-		std::shared_ptr<const IPrim>	hit		= nullptr;									///< Pointer to currently closest primitive
```
Comment (ereator) [inline modules/core/Ray.h]: Is it possible to leave it const ?

File: modules/core/PrimTriangle.h [hunk 3]
```diff
@@ -44,12 +44,12 @@ namespace rt {
 		{}
 		DllExport virtual ~CPrimTriangle(void) = default;
 		
-		DllExport virtual bool	intersect(Ray& ray) const override;
```
Comment (ereator) [inline modules/core/PrimTriangle.h]: is it possible to leave it as const ?

File: modules/core/PrimTriangle.cpp [hunk 4]
```diff
@@ -3,7 +3,7 @@
 #include "Transform.h"
 
 namespace rt {
-	bool CPrimTriangle::intersect(Ray& ray) const
```
Comment (ereator) [inline modules/core/PrimTriangle.cpp]: is it possible to leave it as const ?

File: modules/core/PrimSphere.h [hunk 5]
```diff
@@ -27,11 +27,11 @@ namespace rt {
 		{}
 		DllExport virtual ~CPrimSphere(void) = default;
 
-		DllExport virtual bool 			intersect(Ray& ray) const override;
```
Comment (ereator) [inline modules/core/PrimSphere.h]: is it possible to leave it as const ?

File: modules/core/PrimSphere.cpp [hunk 6]
```diff
@@ -4,7 +4,7 @@
 #include "macroses.h"
 
 namespace rt {
-	bool CPrimSphere::intersect(Ray& ray) const
```
Comment (ereator) [inline modules/core/PrimSphere.cpp]: is it possible to leave it as const ?

File: modules/core/PrimSphere.cpp [hunk 7]
```diff
@@ -56,14 +56,14 @@ namespace rt {
 	/// @todo Optimize it
 	bool CPrimSphere::if_intersect(const Ray& ray) const
 	{
-		return intersect(lvalue_cast(Ray(ray)));
-	}
+        return const_cast<CPrimSphere*>(this)->intersect(lvalue_cast(Ray(ray)));
```
Comment (ereator) [inline modules/core/PrimSphere.cpp]: This casting is not quite clear. if there are cases when `this` is not CPromSphere?

File: modules/core/PrimPlane.h [hunk 8]
```diff
@@ -29,11 +29,11 @@ namespace rt {
 		}
 		DllExport virtual ~CPrimPlane(void) = default;
 
-		DllExport virtual bool 			intersect(Ray& ray) const override;
```
Comment (ereator) [inline modules/core/PrimPlane.h]: is it possible to leave it as const ?

File: modules/core/PrimPlane.cpp [hunk 9]
```diff
@@ -3,7 +3,7 @@
 #include "Transform.h"
 
 namespace rt {
-	bool CPrimPlane::intersect(Ray& ray) const
```
Comment (ereator) [inline modules/core/PrimPlane.cpp]: is it possible to leave it as const ?

File: modules/core/PrimDisc.h [hunk 10]
```diff
@@ -29,11 +29,11 @@ namespace rt {
 		{}
 		DllExport virtual ~CPrimDisc(void) = default;
 		
-		DllExport virtual bool				intersect(Ray& ray) const override;
```
Comment (ereator) [inline modules/core/PrimDisc.h]: is it possible to leave it as const ?

File: modules/core/PrimDisc.cpp [hunk 11]
```diff
@@ -3,7 +3,7 @@
 #include "Transform.h"
 
 namespace rt {
-	bool CPrimDisc::intersect(Ray& ray) const
```
Comment (ereator) [inline modules/core/PrimDisc.cpp]: is it possible to leave it as const ?

File: modules/core/CompositeGeometry.cpp [hunk 12]
```diff
@@ -28,7 +27,7 @@ namespace rt {
 #endif
     }
 
-    bool CCompositeGeometry::intersect(Ray &ray) const {
```
Comment (ereator) [inline modules/core/CompositeGeometry.cpp]: is it possible to leave it as const ?

**Comment by ereator** [comment]
I would try not to make the const methods and pointer in ray to be non-const. Let us first concentrate on inverting normals and then on removing the dummy primitive

**Pull request merged**